pilot/pkg/trustbundle/trustbundle.go

```diff
@@ -238,18 +261,28 @@ func (tb *TrustBundle) fetchRemoteTrustAnchors() {
 	remoteCerts := []string{}
 
 	currentTrustDomain := spiffe.GetTrustDomain()
-	for _, endpoint := range remoteEndpoints {
+
+	for endpointURI, remoteTrustAnchor := range remoteEndpoints {
 		trustDomainAnchorMap, err := spiffe.RetrieveSpiffeBundleRootCerts(
-			map[string]string{currentTrustDomain: endpoint}, tb.remoteCaCertPool, remoteTimeout)
-		if err != nil {
-			trustBundleLog.Errorf("unable to fetch trust Anchors from endpoint %s: %s", endpoint, err)
-			continue
-		}
-		certs := trustDomainAnchorMap[currentTrustDomain]
-		for _, cert := range certs {
-			certStr := string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}))
-			trustBundleLog.Debugf("from endpoint %v, fetched trust anchor cert: %v", endpoint, certStr)
-			remoteCerts = append(remoteCerts, certStr)
+			map[string]string{currentTrustDomain: endpointURI}, tb.remoteCaCertPool, remoteTimeout)
+		if err == nil {
+			remoteTrustAnchor.PurgeCerts()
+			certs := trustDomainAnchorMap[currentTrustDomain]
+			for _, cert := range certs {
+				certStr := string(pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: cert.Raw}))
+				trustBundleLog.Debugf("from endpoint %v, fetched trust anchor cert: %v", endpointURI, certStr)
+				remoteCerts = append(remoteCerts, certStr)
+				remoteTrustAnchor.AddCert(certStr)
+			}
+		} else {
+			if !remoteEndpoints[endpointURI].IncrementRetryOrFail() {
+				remoteTrustAnchor.PurgeCerts()
```

Comment:
So after 1 day we purge it? This feels a bit odd to me. It seems like its either too high or too low. Either we need to purge the certs when the endpoint is down, in which case we ought to do it sooner than a day, or we don't need to in which case we never should?

Reply:
Not sure I agree. There is clearly a need to delay removal of certs (in the event that an endpoint becomes temporarily unavailable). However, I am equally concerned about having a trustAnchor in our cluster long after an endpoint becomes unreachable. What if the user is unaware of this and doesn't change meshConfig to remove this endpoint?

Reply:
I understand the concern - I think the tricky thing is there is no "good" solution, just the least bad.

A similar issue is leaving "dead" multicluster configs around - I don't think its right for us to prune them. We should make it *very* clear to users that something is wrong, with logs+metrics, but I think removing it is a bit too far.

Especially when we have some logic that will be exercised only in a day, it seems concerning that there is such a huge gap between something going wrong and the impact of that